### ISSUE TITLE
ICE with OsStr::new in match clause

### ISSUE THREAD

**Issue by ColonelThirtyTwo** [rust-lang/rust#56387]
Example code:

```rust
use std::ffi::OsStr;

fn main() {
	let s = None;
	match s {
		Some(OsStr::new("png")) => {},
	};
}

```

Building:

```plain
> env RUST_BACKTRACE=1 cargo build
   Compiling rust-crash v0.1.0 (/home/col/Downloads/rust-crash)
thread 'main' panicked at 'expected fn type', libcore/option.rs:1000:5                        
stack backtrace:                                                                              
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace                               
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49                                 
   1: std::sys_common::backtrace::print                                                       
             at libstd/sys_common/backtrace.rs:71                                             
             at libstd/sys_common/backtrace.rs:59                                             
   2: std::panicking::default_hook::{{closure}}                                               
             at libstd/panicking.rs:211                                                       
   3: std::panicking::default_hook                                                            
             at libstd/panicking.rs:227                                                       
   4: rustc::util::common::panic_hook                                                         
   5: std::panicking::rust_panic_with_hook                                                    
             at libstd/panicking.rs:481                                                       
   6: std::panicking::continue_panic_fmt                                                      
             at libstd/panicking.rs:391                                                       
   7: rust_begin_unwind                                                                       
             at libstd/panicking.rs:326                                                       
   8: core::panicking::panic_fmt                                                              
             at libcore/panicking.rs:77                                                       
   9: core::option::expect_failed                                                             
             at libcore/option.rs:1000                                                        
  10: rustc_typeck::check::_match::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, 'tcx>>::check_pat_tuple_struct
  11: rustc_typeck::check::_match::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, 'tcx>>::check_pat_walk
  12: rustc_typeck::check::_match::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, 'tcx>>::check_pat_tuple_struct
  13: rustc_typeck::check::_match::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, 'tcx>>::check_pat_walk
  14: <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T, I>>::from_iter                         
  15: rustc_typeck::check::FnCtxt::check_expr_kind                                            
  16: rustc_typeck::check::FnCtxt::check_expr_with_expectation_and_needs                      
  17: rustc_typeck::check::FnCtxt::check_block_with_expected                                  
  18: rustc_typeck::check::FnCtxt::check_expr_kind                                            
  19: rustc_typeck::check::FnCtxt::check_expr_with_expectation_and_needs                      
  20: rustc_typeck::check::FnCtxt::check_return_expr                                          
  21: rustc_typeck::check::check_fn                                                           
  22: rustc::ty::context::tls::with_related_context                                           
  23: rustc::infer::InferCtxtBuilder::enter                                                   
  24: rustc_typeck::check::typeck_tables_of                                                   
  25: rustc::ty::query::__query_compute::typeck_tables_of                                     
  26: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::typeck_tables_of<'tcx>>::compute
  27: rustc::ty::context::tls::with_context                                                   
  28: rustc::dep_graph::graph::DepGraph::with_task_impl                                       
  29: rustc::ty::context::tls::with_related_context                                           
  30: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  31: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  32: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::ensure_query
  33: rustc_typeck::check::typeck_item_bodies                                                 
  34: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::typeck_item_bodies<'tcx>>::compute
  35: rustc::ty::context::tls::with_context                                                   
  36: rustc::dep_graph::graph::DepGraph::with_task_impl                                       
  37: rustc::ty::context::tls::with_related_context                                           
  38: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  39: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  40: rustc_typeck::check_crate                                                               
  41: rustc::ty::context::tls::enter_context                                                  
  42: <std::thread::local::LocalKey<T>>::with                                                 
  43: rustc::ty::context::TyCtxt::create_and_enter                                            
  44: rustc_driver::driver::compile_input                                                     
  45: rustc_driver::run_compiler_with_pool                                                    
  46: rustc_driver::driver::spawn_thread_pool                                                 
  47: rustc_driver::run_compiler                                                              
  48: <scoped_tls::ScopedKey<T>>::set                                                         
  49: syntax::with_globals                                                                    
  50: __rust_maybe_catch_panic                                                                
             at libpanic_unwind/lib.rs:103                                                    
  51: rustc_driver::run                                                                       
  52: rustc_driver::main                                                                      
  53: std::rt::lang_start::{{closure}}                                                        
  54: std::panicking::try::do_call                                                            
             at libstd/rt.rs:59                                                               
             at libstd/panicking.rs:310                                                       
  55: __rust_maybe_catch_panic                                                                
             at libpanic_unwind/lib.rs:103                                                    
  56: std::rt::lang_start_internal                                                            
             at libstd/panicking.rs:289                                                       
             at libstd/panic.rs:392                                                           
             at libstd/rt.rs:58                                                               
  57: main                                                                                    
  58: __libc_start_main                                                                       
  59: <unknown>                                                                               
query stack during panic:                                                                     
#0 [typeck_tables_of] processing `main`                                                       
#1 [typeck_item_bodies] type-checking all item bodies                                         
end of query stack                                                                            
                                                                                              
error: internal compiler error: unexpected panic                                              
                                                                                              
note: the compiler unexpectedly panicked. this is a bug.                                      
                                                                                              
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
                                                                                              
note: rustc 1.30.1 (1433507eb 2018-11-07) running on x86_64-unknown-linux-gnu                 
                                                                                              
note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin                          
                                                                                              
note: some of the compiler flags provided by cargo are hidden                                 
                                                                                              
error: Could not compile `rust-crash`.                                                        

To learn more, run the command again with --verbose.
```

**Comment by ColonelThirtyTwo** [comment]
Slightly smaller case:

```rust

use std::ffi::OsStr;

fn main() {
	let s = OsStr::new("asdf");
	match s {
		OsStr::new("png") => {},
	};
}
```

**Comment by bombless** [comment]
Cannot reproduce from 1.31 [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f674c966a937d3c37ca2835f203a2fb3)

**Comment by ColonelThirtyTwo** [comment]
Yeah looks like its fixed in 1.31
